From ce77c266a629354738a6e363cd696fd58499491b Mon Sep 17 00:00:00 2001
From: Phoenix616 <mail@moep.tv>
Date: Wed, 15 May 2019 20:06:44 +0100
Subject: [PATCH] Only grow cactus/chorus fruit on random tick (MC-113809)


diff --git a/src/main/java/net/minecraft/server/Block.java b/src/main/java/net/minecraft/server/Block.java
index 2e460a7f..371b6962 100644
--- a/src/main/java/net/minecraft/server/Block.java
+++ b/src/main/java/net/minecraft/server/Block.java
@@ -22,6 +22,7 @@ public class Block implements IMaterial {
     protected final boolean i;
     protected final SoundEffectType stepSound;
     protected final Material material;
+    public boolean randomTick = false; // Phoenix - fix MC-113809
     // Paper start
     public co.aikar.timings.Timing timing;
     public co.aikar.timings.Timing getTiming() {
diff --git a/src/main/java/net/minecraft/server/BlockCactus.java b/src/main/java/net/minecraft/server/BlockCactus.java
index b6806d8d..53330d9b 100644
--- a/src/main/java/net/minecraft/server/BlockCactus.java
+++ b/src/main/java/net/minecraft/server/BlockCactus.java
@@ -20,6 +20,7 @@ public class BlockCactus extends Block {
         if (!iblockdata.canPlace(world, blockposition)) {
             world.setAir(blockposition, true);
         } else {
+            if (!randomTick) return; // Phoenix - fix MC-113809
             BlockPosition blockposition1 = blockposition.up();
 
             if (world.isEmpty(blockposition1)) {
diff --git a/src/main/java/net/minecraft/server/BlockChorusFlower.java b/src/main/java/net/minecraft/server/BlockChorusFlower.java
index 5063c94c..ea6ab294 100644
--- a/src/main/java/net/minecraft/server/BlockChorusFlower.java
+++ b/src/main/java/net/minecraft/server/BlockChorusFlower.java
@@ -25,6 +25,7 @@ public class BlockChorusFlower extends Block {
         if (!iblockdata.canPlace(world, blockposition)) {
             world.setAir(blockposition, true);
         } else {
+            if (!randomTick) return; // Phoenix - fix MC-113809
             BlockPosition blockposition1 = blockposition.up();
 
             if (world.isEmpty(blockposition1) && blockposition1.getY() < 256) {
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index ee071ba2..3cfd30e6 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -549,7 +549,9 @@ public class WorldServer extends World implements IAsyncTaskHandler {
 
                                 this.methodProfiler.enter("randomTick");
                                 if (iblockdata.t()) {
+                                    iblockdata.getBlock().randomTick = true; // Phoenix - fix MC-113809
                                     iblockdata.b((World) this, new BlockPosition(i2 + j, k2 + chunksection.getYPosition(), j2 + k), this.random);
+                                    iblockdata.getBlock().randomTick = false; // Phoenix - fix MC-113809
                                 }
 
                                 if (fluid.h()) {
-- 
2.18.0.windows.1

